populator-machinery/controller.go

```diff
@@ -108,13 +116,92 @@ type controller struct {
+	Kubeconfig   string
+	HttpEndpoint string
+	MetricsPath  string
+	// Namespace is the namespace that all populator resources run in
+	Namespace string
+	Prefix    string
+	// Gk is the GroupKind for the CR data source created by user
+	Gk schema.GroupKind
+	// Gvr is the GroupVersionResource for the CR data source created by user
+	Gvr schema.GroupVersionResource
+	// PodConfig is the configuration for creating populator pod. Either PodConfig or ProviderFunctionConfig should
+	// be specified. PodConfig and ProviderFunctionConfig can't be provided at the same time
+	PodConfig *PodConfig
+	// ProviderFunctionConfig is the configuration for invoking provider functions. Either PodConfig or ProviderFunctionConfig should
+	// be specified. PodConfig and ProviderFunctionConfig can't be provided at the same time
+	ProviderFunctionConfig *ProviderFunctionConfig
+}
+
+type PodConfig struct {
+	// ImageName is the container image name
+	ImageName string
+	// MountPath is the path inside of the container at which the volume should be mounted
+	MountPath string
+	// DevicePath is the path inside of the container that the device will be mapped to
+	DevicePath string
+	// PopulatorArgs is the args for the populator pod
+	PopulatorArgs func(bool, *unstructured.Unstructured) ([]string, error)
+}
+
+type ProviderFunctionConfig struct {
+	// PopulateFn is the provider specific data population function
+	PopulateFn func(context.Context, PopulatorParams) error
+	// PopulateCompleteFn is the provider specific data population completeness check function, return true when data transfer gets completed
+	PopulateCompleteFn func(context.Context, PopulatorParams) (bool, error)
+	// PopulateCleanupFn is the provider specific data population cleanup function, cleanup resouces after data population completed
+	PopulateCleanupFn func(context.Context, PopulatorParams) error
+}
+
+// PopulatorParams includes the parameters passing to provider specific functions
+type PopulatorParams struct {
+	KubeClient kubernetes.Interface
+	// Pvc is the original PVC created by user
+	Pvc *corev1.PersistentVolumeClaim
+	// PvcPrime is the temporary PVC created by volume populator
+	PvcPrime *corev1.PersistentVolumeClaim
+	// StorageClass is the original StorageClass Pvc refer to
+	StorageClass *storagev1.StorageClass
+	// Unstructured is the CR data source created by user
+	Unstructured *unstructured.Unstructured
 }
 
 func RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
 	gk schema.GroupKind, gvr schema.GroupVersionResource, mountPath, devicePath string,
 	populatorArgs func(bool, *unstructured.Unstructured) ([]string, error),
 ) {
-	klog.Infof("Starting populator controller for %s", gk)
+	podConfig := &PodConfig{
+		ImageName:     imageName,
+		MountPath:     mountPath,
+		DevicePath:    devicePath,
+		PopulatorArgs: populatorArgs,
+	}
+	vpcfg := &VolumePopulatorConfig{
```

Comment:
This does not need to be a pointer anymore

Reply:
Updated